README
# README for the itrans package, 5.32

                    Version 5.32, of the itrans package
     (Supports Devanagari, Gujarati, Telugu, Kannada, Bengali, Tamil,
                    Punjabi, and Romanized Sanskrit)
            -----------------------------------------------------
                    http://www.aczoom.com/itrans/
            -----------------------------------------------------

This is a package for printing text in Indian language scripts.

This package only does the transliteration mapping, the fonts
may be developed elsewhere.

ITRANS is distributed in the following archives:

    [required]
    itransNN.zip  - ITRANS source code (Unix format) in a ZIP archive
                      - also includes binary for MS-DOS
                  - includes docs in .itx (ITRANS) format
    itransfn.zip  - fonts for use with ITRANS
    iupdateN.zip  - Updates to itransNN.zip [when needed]

    [pre-compiled binary archives, includes fonts for use with ITRANS]
    itransNN-i386.tgz   - for i386 Linux platforms (tested on Slackware)
    itransNN-win32.zip  - for Win32 (Windows 95/NT or newer)

    [optional]
    itransht.zip  - docs in HTML (using Ross Moore's Latex2HTML for ITRANS)
    itransps.zip  - docs in PostScript format, for printing.

----- Navigating the complexities of ITRANS:
 Beginning users: Consult these files: idoc.itx and other *.itx files in the
 doc/ directory, and if you have to deal with installation issues, also
 check out INSTALL.unix or INSTALL.pc in the top directory.

Also, check out the ITRANS home page for more info and links to the
Online ITRANS interface and other e-mail tools that make using ITRANS
much easier, no installation required:

        http://www.aczoom.com/itrans/

Previous release ITRANS users: See the CHANGES file for list of changes.

-----

Directories & Files (all directories have README files with more info):

src/            ITRANS source code, and makefiles.
                Also includes binary for Linux (ELF) and MS-DOS.

INSTALL.unix    Installation instructions for ITRANS on Unix systems.
INSTALL.pc      Installation instructions for ITRANS on DOS/Win systems.

TRANS.TXT       Quick review of the ITRANS encoding table.

CHANGES         list of changes between consecutive released ITRANS versions.

lib/fonts       All fonts used by ITRANS.

lib/fonts/README    List all font files in lib/fonts and their uses.

itrans.lst, itransps.lst, itransht.lst, itransfn.lst
		List of files in the respective ITRANS archives

lib/            Input files that define char composition - the *.ifm files.
                Also contains examples of environment variables needed,
                dvipsrc scripts, etc.

doc/            ITRANS documentation.

doc/idoc.itx    The main ITRANS manual
		(all docs also available in PostScript and HTML formats).

doc/ex_*.itx    Example input files, for LaTeX, HTML, Unicode, etc output.

contrib/        Additional files and tools that may be useful.

----
    COPYRIGHT file has licensing information - Simplified BSD License.
----
Avinash Chopde
e-mail: avinash@aczoom.com
home page: http://www.aczoom.com/

---------------------------------------------------------------------


CHANGES
--------------------------
****** Version 5.34, 10 Apr 2013
    1) Have itrans automatically pass through extended Unicode (non-ASCII) UTF-8 chars
       from the input. There was a problem in that CSX codes like 0o342 (0xE2) for AA end up
       breaking up Unicode double quote (0xE2 0x80 0x9C) and causing invalid UTF-8 output.

    2) Suppress "Note... echoing to output" messages from itrans output, unless -v is provided.
        "Note (around line %d): character [%c] (dec %d) unrecognized by lexer, echoing to output\n"

    IMPORTANT: Windows is no longer supported. Last pre-built version it has is ITRANS 3.2.

****** Version 5.33, 10 Mar 2013
    1) Suppress copyright messages to stdout. This supports use as a filter,
       so only actual conversion is output to stdout or to output file.

****** Version 5.32, 06 Feb 2011, 09 Dec 2012
    1) [06 Feb 2011] Discovered a problem with UTF-8 output for Kannada, Telugu. The
       ukan.ifm etc files need to be processed like Devanagari, i.e,
       C+C+V and not C+V+C-Viraam which is done for TeX output.
    2) [Dec 2012]
       * UTF-8 Roman output for j~n fixed. Needs to use character code 241 in UTF,
       and 164 in ISO-8859-1, or extended ASCII. jnh and gya changed in uroman.ifm.
       * "La" (lda) also changed - now is l-underbar in UTF-8 (but not ISO-8859) output.
       tblall docs updated: http://www.aczoom.com/itrans/html/tblall/node4.html
       though there may be errors in conversion due to invalid Latex2HTML setup.

****** Version 5.31 (executables still 5.30), 01 Dec 2009
    1) This is primarily a packaging change, with some bug fixes in config
       files. No changes in the executable.

    2) dvng: instead of using the dvng font, it now uses dvnb which contains
       the GY j~n character needed.
       The current TeX distribution, texlive, has latest Velthuis Devnag
       package which does not have j~n, and also removes old \char43.
       The "dvnb" font has j~n at 224 and 236. So, updated all tex files
       to refer to dvnb10 instead of dvng10 (though kept names \largedvng, etc
       the same), and change ja-nya and gya in dvng and dvngfull IFM files
       to refer to 224/236.
       Updated files:
       idevn.tex
       webincl.tex
       dvng.ifm
       dvngfull.ifm

       Note that texlive may have an older version of idevn.tex, so
       make sure the itrans version of idevn.tex is picked up first by 
       LaTeX, for example, by adding these lines to appropriate
       environment variables or to /usr/local/texlive/2009/texmf.cnf
       (see file INSTALL.unix)

    3) The web input file webincl.tex is now available in this package
       lib/ directory. It is used by the itrans cgi script, which is
       available in itrans-cgi.zip archive.

    4) Docs
       .ps files removed, now using .pdf files instead.
       .html files regenerated, managed to get latex2html running.
       See doc/README for some details, as well as contrib/mkl2h

****** Version 5.30, 22 July 2001
    1) Added support for Unicode UTF-8 output
    2) Binary pre-built installations created for Linux-i386 and Win32
    3) Created support for "make install" to install itrans on
       a Unix box (Linux, etc).

    ** 10 Dec, 2001
    1) Minor fixes to makefile.gcc, etc - "make install" should now work

    ** 5 Oct, 2002
    1) itrans.h mods for MacOS X/Darwin compilcation, from Shrisha Rao

    ** 12 Aug, 2003
    1) Updated Gurmukhi doc to make it clear the tables have #usecsx on.

    ** 22 Feb, 2003
    1) Updated ubeng.ifm, added Unicode yya (ya-dot)

    ** 29 Jul, 2006
    1) buffer checks in iyacc.y and dflt_mkl.c - handle bad input, too many
    half-consonants in a letter - as in when users send in binary files
    instead of text to the Online Web Interface to ITRANS.
    2) removed: itrans16.exe  - 16-bit Windows exe, for Windows 3.1,
    no longer supported. 32-bit Windows itrans.exe is available.

    ** 28 Apr, 2007
    1) idevn.tex - add % after font definition \normaldvng etc, to
    avoid a space character after every #indian or ##

    ** 26 Mar, 2007
    1) uroman.ifm:
    fixed tmpl-ii should be 0x012B was 0x16B same as tmpl-uu.
    2) udvng.ifm:
    make O be the VOWEL SIGN CANDRA O  (AA + ardha-chandra)
    Only Unicode conversion supports this, for now.

    ** 27 Sep, 2009
    1) Email, Web updates in readme
    2) tlgutx.ifm sa was too like na, created a new tmplC for sa.
    3) The current TeX distribution, texlive, has latest Velthuis Devnag
       package which does not have j~n, and also removes old \char43.
       The "dvnb" font has j~n at 224 and 236. So, updated all tex files
       to refer to dvnb10 instead of dvng10 (though kept names \largedvng, etc
       the same), and change ja-nya and gya in dvng and dvngfull IFM files
       to refer to 224/236.

--------------------------
****** Version 5.22, 18 December 1998 [also March 1999, and Feb 2000]
Minor changes to 5.20
    1) Fixed bug in Telugu processing - code was ignoring explicit IFM
       rules for multiple consonants (C1-C2 : ka-kha, etc).
    2) Added Tikkana.ifm for TrueType telugu font (Potana, Tikkana, etc).
    3) Minor:
       - Added pagenumbering in Devanagari when
         \usepackage[devnag]{itrans} used in input files [NOT WORKING] 
       - xdvng.ifm now has ka-shha rule (== ksha)
       - contrib/latex2html.txt updated, all html docs regenerated,
         many had missing images; beng.itx changes for latex2html [March 1999]
       - HTML output font commands now accept + in SIZE spec, to allow
         #<langfont>=<FONT FACE="xdvng" SIZE="+3">

--------------------------
****** Version 5.20, 16 September 1998
This release contains additions and bug fixes only. All ITRANS 5.1
documents will be correctly handled by this release.
    1) Added additional ITRANS codes:
       RRi == R^i     RRI == R^I
       LLi == L^i     LLI == L^I
       '   == .a      (only when #usecsx is in effect)
    2) Gurmukhi now supported.

src/imap.h
#ifndef IMAP_H
#define IMAP_H

/*
 *========================================================================== 
 * Copyright 1991-1992 Avinash Chopde, All Rights Reserved.
 */

/* static char S_RCSID[] = "$Header: /home/cvsroot/itrans/nextrel/src/imap.h,v 1.3 1998/09/13 00:38:46 avinash Exp $"; */

#include "y_tab.h"
	/* not including y.tab.h, see makefile which renames it to
	 * y_tab.h for compatibility with DOS machines
	 */

#define	CNAMELEN	16
	/* max length of the code name of any indian char */

typedef struct {
    int		token; /* char token - assigned by y_tab.h */
    char	codename[CNAMELEN+1]; /* name assigned to it in ifm file */
} ifm_enc_t;

/* Some arbit representation of input token names, this is pretty
 * much free-form - no duplicate names allowed.
 * Used to parse the Devn Metrics file - see font.c
 * XXX
 * Will presumably be used later, to make error messages a bit
 * descriptive as to what actually expected...

 * Note:  names should be letters only (case is significant)
 *        do not use "-" in the names.
 * Any changes here implies that the file font.c needs to be edited:
 * the struct S_ifm_map has to be filled in accordingly.
 */

#define A_CNAME		"a"
#define AA_CNAME	"aa"
#define I_CNAME		"i"
#define II_CNAME	"ii"
#define U_CNAME		"u"
#define UU_CNAME	"uu"
#define RI_CNAME	"ri"
#define RII_CNAME	"rii"
#define LI_CNAME	"li"
#define LII_CNAME	"lii"
#define AY_CNAME	"ay"
#define AAY_CNAME	"aay"
#define AI_CNAME	"ai"
#define O_CNAME		"o"
#define OO_CNAME	"oo"
#define AU_CNAME	"au"
#define AM_CNAME	"am"
#define AHA_CNAME	"aha"
#define HALF_CNAME	"half"
#define IMPLICIT_CNAME	"implicit"
#define CONJUNCT_CNAME	"conjunct"
#define KA_CNAME	"ka"
#define KADOT_CNAME	"kadot"
#define KHA_CNAME	"kha"
#define KHADOT_CNAME	"khadot"
#define GA_CNAME	"ga"
#define GADOT_CNAME	"gadot"
#define GHA_CNAME	"gha"
#define NGA_CNAME	"nga"
#define CHA_CNAME	"cha"
#define CHHA_CNAME	"chha"
#define JA_CNAME	"ja"
#define JADOT_CNAME	"jadot"
#define JHA_CNAME	"jha"
#define JNH_CNAME	"jnh"
#define TTA_CNAME	"tta"
#define TTHA_CNAME	"ttha"
#define DDA_CNAME	"dda"
#define DDADOT_CNAME	"ddadot"
#define DDHA_CNAME	"ddha"
#define DDHADOT_CNAME	"ddhadot"
#define NNA_CNAME	"nna"
#define TA_CNAME	"ta"
#define THA_CNAME	"tha"
#define DA_CNAME	"da"
#define DHA_CNAME	"dha"
#define NA_CNAME	"na"
#define PA_CNAME	"pa"
#define PHA_CNAME	"pha"
#define PHADOT_CNAME	"phadot"
#define BA_CNAME	"ba"
#define BHA_CNAME	"bha"
#define MA_CNAME	"ma"
#define YA_CNAME	"ya"
#define YYA_CNAME	"yya"
#define RA_CNAME	"ra"
#define LA_CNAME	"la"
#define VA_CNAME	"va"
#define SHA_CNAME	"sha"
#define SHHA_CNAME	"shha"
#define SA_CNAME	"sa"
#define HA_CNAME	"ha"
#define LDA_CNAME	"lda"
#define KSHA_CNAME	"ksha"
#define GYA_CNAME	"gya"
#define NYA_CNAME	"nya"
#define NNX_CNAME	"nnx"
#define RRA_CNAME	"rra"
#define RHA_CNAME	"rha"

#define RA_HALF_CNAME	"rahalf"
#define ANUSVARA_CNAME	"anusvara"
#define CHANDRA_CNAME	"chandra"
#define CHANDRA_BN_CNAME	"chandrabindu"
#define VIRAAM_CNAME	"viraam"
#define AVAGRAHA_CNAME	"avagraha"
#define SRI_CNAME	"sri"
#define AUM_CNAME	"aum"

/* miscellaneous characters -- these are not parsed by ilex.l or
 * iyacc.y, they are simply passed on to process_otherchar() in itrans.c
 * But, they may still have definitions  in the IFM file, used mainly
 * to take care of fonts where the digits (numbers) are not in the
 * correct ASCII location.
 * Currently, only digits may be defined in the IFM file.
 * These are not in the S_ifm_map table, so do not affect any other
 * piece of code (unlike the KA_CNAME, etc codes above)
 */
#define ZERO_CNAME	"zero"
#define ONE_CNAME	"one"
#define TWO_CNAME	"two"
#define THREE_CNAME	"three"
#define FOUR_CNAME	"four"
#define FIVE_CNAME	"five"
#define SIX_CNAME	"six"
#define SEVEN_CNAME	"seven"
#define EIGHT_CNAME	"eight"
#define NINE_CNAME	"nine"

/* num indian lang characters */
/* see font.c also */
#define FIRSTCHAR	_I_(A_TOK)
#define LASTCHAR	_I_(AUM_TOK)
#define	NUMCHARS	(LASTCHAR-FIRSTCHAR+1)
#define FIRSTVOW	_I_(A_TOK)
#define LASTVOW		_I_(AHA_TOK)
#define FIRSTCONS	_I_(KA_TOK)
#define LASTCONS	_I_(PHADOT_TOK)
	/* allow for extra chars that the user may define */
#define NUMEXTRA	32
#define	NUMKHADI	(NUMCHARS+NUMEXTRA)
#define	OFFSET_TOK	(DUMMY_TOK+1)

/* num postscript chars - 8 bits = 0 to 255 */
#define	NUMPSCHARS	256

#endif /* IMAP_H */
